### ISSUE TITLE
Unexpected 4.08 Request Entity Incomplete reponse with block1

### ISSUE THREAD

**Issue by sbernard31** [eclipse/californium#537]
We faced this kind of error after a while using block1 with same IP/port same URI with cf 2.0.0-M6 (reproducible with 2.0.0-M7).

See capture here : 
```
     27 17.921800      64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:3915, 2.31 Continue, TKN:4b 0f 74 b1, Block #0, /push
     28 17.923302      2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     163    CON, MID:3916, POST, TKN:4c 0f 75 b1, End of Block #1, /push?ep=351711090003168
     29 18.177398      64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:3916, 2.01 Created, TKN:4c 0f 75 b1, End of Block #1, /push
     30 22.460344      2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     260    CON, MID:3917, POST, TKN:4d 0f 79 b1, Block #0, /push?ep=351711090003168
     31 22.682601      64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:3917, 2.31 Continue, TKN:4d 0f 79 b1, Block #0, /push
     32 22.683741      2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     163    CON, MID:3918, POST, TKN:4e 0f 79 b1, End of Block #1, /push?ep=351711090003168
     33 22.958042      64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:3918, 2.01 Created, TKN:4e 0f 79 b1, End of Block #1, /push
     34 27.459536      2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     260    CON, MID:3919, POST, TKN:4f 0f 7e b1, Block #0, /push?ep=351711090003168
     35 27.717607      64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:3919, 2.31 Continue, TKN:4f 0f 7e b1, Block #0, /push
     36 27.718688      2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     162    CON, MID:3920, POST, TKN:50 0f 7e b1, End of Block #1, /push?ep=351711090003168
     37 27.977879      64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:3920, 2.01 Created, TKN:50 0f 7e b1, End of Block #1, /push
     38 32.461906      2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     260    CON, MID:3921, POST, TKN:51 0f 83 b1, Block #0, /push?ep=351711090003168
     39 32.719572      64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:3921, 2.31 Continue, TKN:51 0f 83 b1, Block #0, /push
     40 32.720789      2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     163    CON, MID:3922, POST, TKN:52 0f 83 b1, End of Block #1, /push?ep=351711090003168
     41 32.997208      64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:3922, 2.01 Created, TKN:52 0f 83 b1, End of Block #1, /push
.... .... ... ~ 5 minutes later ... ... ...
    270 322.459758     2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     260    CON, MID:4037, POST, TKN:c5 0f a5 b2, Block #0, /push?ep=351711090003168
    271 322.718599     64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:4037, 2.31 Continue, TKN:c5 0f a5 b2, Block #0, /push
    272 322.719724     2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     163    CON, MID:4038, POST, TKN:c6 0f a5 b2, End of Block #1, /push?ep=351711090003168
    273 322.936894     64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     123    ACK, MID:4038, 4.08 Request Entity Incomplete, TKN:c6 0f a5 b2, End of Block #1, /push (text/plain)
    274 327.460062     2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     260    CON, MID:4039, POST, TKN:c7 0f aa b2, Block #0, /push?ep=351711090003168
    275 327.698077     64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:4039, 2.31 Continue, TKN:c7 0f aa b2, Block #0, /push
    276 327.699454     2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     164    CON, MID:4040, POST, TKN:c8 0f aa b2, End of Block #1, /push?ep=351711090003168
    277 327.961424     64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:4040, 2.01 Created, TKN:c8 0f aa b2, End of Block #1, /push
    278 332.461631     2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     260    CON, MID:4041, POST, TKN:c9 0f af b2, Block #0, /push?ep=351711090003168
    279 332.698640     64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:4041, 2.31 Continue, TKN:c9 0f af b2, Block #0, /push
    280 332.699763     2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     163    CON, MID:4042, POST, TKN:ca 0f af b2, End of Block #1, /push?ep=351711090003168
    281 332.958887     64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:4042, 2.01 Created, TKN:ca 0f af b2, End of Block #1, /push
    282 337.460227     2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     260    CON, MID:4043, POST, TKN:cb 0f b4 b2, Block #0, /push?ep=351711090003168
    283 337.698859     64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     104    ACK, MID:4043, 2.31 Continue, TKN:cb 0f b4 b2, Block #0, /push
    284 337.701265     2001:569:6ff9:6647:7520:b874:b77:b628 64:ff9b::3411:8e3e    56830    5686     CoAP     163    CON, MID:4044, POST, TKN:cc 0f b4 b2, End of Block #1, /push?ep=351711090003168
    285 337.917032     64:ff9b::3411:8e3e    2001:569:6ff9:6647:7520:b874:b77:b628 5686     56830    CoAP     123    ACK, MID:4044, 4.08 Request Entity Incomplete, TKN:cc 0f b4 b2, End of Block #1, /push (text/plain)
```

After investigation, I suspect this is the `CleaningTask` of block status from BlockwiseLayer which is not stopped when we received the last block and so could interfere with next transfer using same URI same IP/port.

Reading the code, I can not find other explanation. The error appears ~ 5 minutes later which is the BLOCKWISE_STATUS_LIFETIME. This confirm the idea.

I will propose a PR for this issue.

**Comment by boaks** [comment]
Seems, that the "cure" for the message exchange house-keeping has it's own house-keeping issues.

Do you see any reason (except of the message exchange house-keeping) for this blockwise status clean up? If there is one, sure your PR fixes this issue.
But if there is no reason left, I would rather go for removing than for fixing it.

**Comment by sbernard31** [comment]
> Do you see any reason (except of the message exchange house-keeping) for this blockwise status clean up? 

1.  house-keeping for blockwiselayer itselft (clearing status)
2.  raise timeout event on request.

> But if there is no reason left, I would rather go for removing than for fixing it.

Silly question but , If we imagine there is no reason except exchange house keeping how do you plan to handle exchange house keeping ?

**Comment by sbernard31** [comment]
> Seems, that the "cure" for the message exchange house-keeping has it's own house-keeping issues.

By the way, I'm not sure what you name "the 'cure' fore the message exchange house-keeping", but this issue was present in californium 2.0.0-M5

**Comment by boaks** [comment]
> If we imagine there is no reason except exchange house keeping how do you plan to handle exchange house keeping ?

By a more "encapsulated/object orientated" implementation. Currently the exchanges and key artifacts handled for blockwise in several cases and as far as I understood the logs, the leak was caused by a "broken" assumption.
 
https://github.com/eclipse/californium/issues/429#issuecomment-361589581

But, yes three is more to work and I'm trying to sort it out.

**Comment by boaks** [comment]
> By the way, I'm not sure what you name "the 'cure' for the message exchange house-keeping", but this issue was present in californium 2.0.0-M5

OK, then I miss-related this issue to PR #441

**Comment by sbernard31** [comment]
Fixed by #538
